File: runtime/perl/units/rt/i32_trunc_u.pl
# requires: rt/trap
use POSIX ();
sub i32_trunc_u {
    my $x = $_[0];
    Rt::trap('invalid conversion to integer') if $x != $x;
    Rt::trap('integer overflow') unless $x > -1.0 && $x < 4294967296.0;
    return POSIX::trunc($x) & 0xFFFFFFFF;
}

File: runtime/java/units/rt/i32_trunc_u.java
// requires: rt/trap
// i32.trunc_f32_u / i32.trunc_f64_u: trapping unsigned truncation.
// The result is the u32 value carried in the low 32 bits of an `int` (e.g. 0xffffffff for
// 4294967295).
// Java's cast saturates, so NaN/out-of-range are checked.
static int i32_trunc_u(double x) {
    if (Double.isNaN(x)) {
        trap("invalid conversion to integer");
    }
    if (Double.isInfinite(x)) {
        trap("integer overflow");
    }
    double t = x < 0 ? Math.ceil(x) : Math.floor(x);
    if (t < 0 || t > 4294967295.0) {
        trap("integer overflow");
    }
    return (int) (long) t;
}
